src/clarity-angular/input-fields/providers/data-mask.ts

```diff
@@ -0,0 +1,72 @@
+/*
+ * Copyright (c) 2016 VMware, Inc. All Rights Reserved.
+ * This software is released under MIT license.
+ * The full license information can be found in LICENSE in the root directory of this project.
+ */
+import {Injectable, Inject} from "@angular/core";
+
+import {Mask} from "../interfaces/mask";
+import {IMaskDefinitionPattern, MASK_PATTERN} from "../interfaces/mask-definition-pattern";
+
+@Injectable()
+export class DataMask implements Mask {
+
+    private EMPTY_CHAR: string = "";
+
+    constructor(@Inject(MASK_PATTERN) private _maskDefinitionPattern: IMaskDefinitionPattern) {}
+
+    applyMask(mask: string, inputValue: string): string {
+        //get the chars from the input field
+        let maskChars = mask.split(this.EMPTY_CHAR);
+        let inputValueChars = inputValue.split(this.EMPTY_CHAR);
+        //get the masked value
+        let maskedValue = this.mapInputWithMaskDefinition(maskChars, inputValueChars);
+
+        return maskedValue;
+    }
+
+    private mapInputWithMaskDefinition(maskChars: string[],
+                                       inputValueChars: string[]): string {
+        let inputValIndex: number = 0;
+        let maskDefnIndex: number = 0;
+        let maskedValue: string[] = [];
+
+        while (inputValIndex < inputValueChars.length &&
+                maskDefnIndex < maskChars.length) {
+            let mask = maskChars[maskDefnIndex];
+            let regexPattern = this._maskDefinitionPattern.maskPattern[mask];
+            let inputChar = inputValueChars[inputValIndex];
+            //if there is a matching regex pattern, then add the char else break
+            if (!regexPattern) {
+                maskedValue.push(mask);
+                maskDefnIndex++;
+                if (mask === inputChar && inputValIndex < inputValueChars.length) {
+                    inputValIndex++;
+                }
+                continue;
+            } else if (regexPattern.test(inputChar)) {
+                maskedValue.push(inputChar);
+                maskDefnIndex++;
+                inputValIndex++;
+                continue;
+            } else {
+                //stop the processing, if the regex does not match
+                break;
```

Comment:
You are missing a graceful error display or at least some feedback to the user when they're typing something wrong (whether it's a wrong character, or continue typing once max length has been reached, ...). Just ignoring use input makes them wonder if the app is not frozen or bugged.